File: 98-validate-binary-search-tree/validate-binary-search-tree.java
class Solution {
    public void inorder(TreeNode root, ArrayList<Integer> ans) {
        if (root == null) return;

        inorder(root.left, ans);

        ans.add(root.val);
        
        inorder(root.right, ans);
    }

    public boolean isValidBST(TreeNode root) {
        ArrayList<Integer> ans = new ArrayList<>();
        inorder(root, ans);

        for (int i = 1; i < ans.size(); i++) {
            if (ans.get(i) <= ans.get(i - 1)) return false;
        }

        return true;
    }
}

File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return validate(root, LONG_MIN, LONG_MAX);
    }

private:
    bool validate(TreeNode* node, long min, long max) {
        if (node == nullptr) return true;

        if (node->val <= min || node->val >= max) return false;

        return validate(node->left, min, node->val) &&
               validate(node->right, node->val, max);
    }
};